Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,379 @@
+		ByID = idIndex.Query
+
+		labelIndex = statedb.Index[Example, string]{
+			Name: "label",
+			FromObject: func(e Example) index.KeySet {
+				return index.NewKeySet(index.String(e.Label))
+			},
+			FromKey: index.String,
+			Unique: false,
+		}
+
+		ByLabel = labelIndex.Query
+	)
+
+	// NewExampleTable creates the table and registers it.
+	func NewExampleTable(db *statedb.DB) (statedb.RWTable[Example], error) {
+		tbl, err := statedb.NewTable(
+			TableName,
+			IDIndex,
+			LabelIndex,
+		)
+		if err != nil {
+			return nil, err
+		}
+		return tbl, db.RegisterTable(tbl)
+	}
+
+	// Cell provides the Table[Example] and registers a controller to populate
+	// the table.
+	var Cell = cell.Module(
+		"example",
+		"Examples"
+
+		// Provide RWTable[Example] privately
+		cell.ProvidePrivate(NewExampleTable),
+
+		// Provide Table[Example] publicly
+		cell.Provide(statedb.RWTable[Example].ToTable),
+
+		// Register a controller that manages the contents of the
+		// table.
+		cell.Invoke(registerExampleController),
+	)
+
+	type exampleController struct {
+		db *statedb.DB
+		examples statedb.RWTable[Example]
+	}
+
+	// loop is a simple control-loop that once a second inserts an example object
+	// with an increasing [ID]. When 100 objects are reached it deletes everything
+	// and starts over.
+	func (e *exampleController) loop(ctx context.Context, health cell.Health) error {
+		id := uint64(0)
+		tick := time.NewTicker(time.Second)
+		defer tick.Stop()
+		for {
+			select {
+			case <-tick.C:
+			case <-ctx.Done():
+				return nil
+			}
+			wtxn := e.db.WriteTxn(examples)
+			if id < 100 {
+				e.examples.Insert(wtxn, Example{ID: id, Label: "hello"})
+			} else {
+				e.examples.DeleteAll(wtxn)
+				id = 0
+			}
+			wtxn.Commit()
+
+			// Report the health of the job. This can be inspected with
+			// "cilium-dbg status --all-health" or with "cilium-dbg statedb health".
+			health.OK(fmt.Sprintf("%d examples inserted", id))
+		}
+	}
+
+	func registerExampleController(jg job.Group, db *statedb.DB, examples statedb.RWTable[Example]) {
+		// Construct the controller and add the loop() method as a one-shot background
+		// job to the module's job group.
+		// When the controller doesn't have any useful API to outside we can use this
+		// pattern instead of "Provide(NewController)" to keep things internal.
+		ctrl := &exampleController{db, examples}
+		jg.Add(job.OneShot(
+			"loop",
+			ctrl.loop,
+		))
+	}
+
+To understand how the table defined by our example module can be consumed, we can
+construct a small mini-application:
+
+.. code-block:: go
+
+	func dumpExamples(jg job.Group, db *statedb.DB, table statedb.Table[examples.Example]) {
+		jg.Add(job.OneShot(
+			"dump",
+			func(ctx context.Context, _ cell.Health) error {
+				// Start tracking changes to the table. This instructs the database
+				// to keep deleted objects off to the side for us to observe.
+				wtxn := db.WriteTxn(table)
+				changes, err := table.Changes(wtxn)
```

Comment:
Any particular reason we need the WriteTxn and ReadTxn is not enough? ( I know the interface signature requires WriteTxn but wondering why?)

txn usage in Changes()
* itxn := txn.getTxn() -> Taken from ReadTxn anyway
* t.Revision(txn) -> which taxes the ReadTxn
* t.LowerBoundWatch(txn..) -> same

Reply:
The reason why `Changes` needs it is that it actually does modify the table:
https://github.com/cilium/statedb/blob/main/txn.go#L265. 

I did at some point think about it internally doing the WriteTxn, but in the end I think this way it's transparent and it's also obvious to the user how this gets sequenced with other WriteTxn's. E.g. This locks the table and we'll see the objects
deleted after we `Commit` this `WriteTxn`.